demos/mqtt/mqtt_demo_connection_sharing/mqtt_demo_connection_sharing.c

```diff
@@ -0,0 +1,1601 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * Demo for showing the use of MQTT APIs to establish an MQTT session,
+ * subscribe to a topic, publish to a topic, receive incoming publishes,
+ * unsubscribe from a topic and disconnect the MQTT session.
+ *
+ * The example is single threaded and uses statically allocated memory;
+ * it uses QOS2 and therefore implements a retransmission mechanism
+ * for Publish messages. Retransmission of publish messages are attempted
+ * when a MQTT connection is established with a session that was already
+ * present. All the outgoing publish messages waiting to receive PUBREC
+ * are resend in this demo. In order to support retransmission all the outgoing
+ * publishes are stored until a PUBREC is received.
+ */
+
+/**
+ * @file mqtt_demo_connection_sharing.c
+ * @brief MQTT Demo application that showcases multiplexing of the same MQTT connection
+ * across multiple topic subscriptions using a subscription manager.
+ */
+
+/* Standard includes. */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* POSIX includes. */
+#include <unistd.h>
+
+/* Include Demo Config as the first non-system header. */
+#include "demo_config.h"
+
+/* MQTT API header. */
+#include "mqtt.h"
+
+/* OpenSSL sockets transport implementation. */
+#include "openssl_posix.h"
```

Comment:
Is it a requirement for this demo to use TLS? It's more straightforward imo for the customer to not have to worry about any certs and just run the demo by providing an endpoint, though I understand if we want more of the demos to use TLS by default